src/hotspot/cpu/x86/c2_MacroAssembler_x86.cpp

```diff
@@ -4740,3 +4740,101 @@ void C2_MacroAssembler::udivmodL(Register rax, Register divisor, Register rdx, R
 }
 #endif
 
+void C2_MacroAssembler::float_class_check(int opcode, Register dst, XMMRegister src, Register temp) {
+
+  int32_t POS_INF = 0x7f800000;
+  int32_t KILL_SIGN_MASK = 0x7fffffff;
+
+  movdl(temp, src);
+  andl(temp, KILL_SIGN_MASK);
+  cmpl(temp, POS_INF);
+  switch (opcode) {
+    case Op_IsFiniteF:
+      setb(Assembler::below, dst);
```

Comment:
This partial write may stall later reads on `dst`, you could emit a `xor dst, dst` before doing the comparison.

Reply:
Will try the `xor dst, dst` and see the performance changes.

Reply:
setb either writes a 0 or 1 based on the condition. So, will it cause partial writes?

Reply:
Actually `setb` only writes the byte portion and leaves the remaining of the register intact, so it would be wrong without clearing the register beforehand.

Reply:
`setb` is producing the correct results and also adding the `xor dst, dst` didn't give any performance improvement. Is it still necessary?

Reply:
> This partial write may stall later reads on `dst`, you could emit a `xor dst, dst` before doing the comparison.

APIs generate boolean result,  and any reader should only be consuming 8 bit result. Thus its not a scenario for partial register stall from HW perspective,  a narrow write followed by wider read cause a partial register stall due to additional cycles penalty incurred while re-assembling the unmodified bits with modified portion.

Reply:
@vamsi-parasa Other boolean producers such as `Conv2B` or `VectorTest` widen the byte value to 32 bit after `setb` so I believe a zeroing is necessary here. You accidentally achieve correct results in tests because the boolean stores only read the least significant 8 bits of the `dst` registers. Other operations such as `test` will read full `int` and may lead to incorrect results.

@jatin-bhateja Boolean is just dressed-up int except in stores so most readers will consume 32 bits of the register.

Thanks.

Reply:
@merykitty You might be right. `setb` was causing a wierd jdk build error in `java.lang.HashMap` due to to error in `isNaN()`. It got fixed by clearing the higher bytes of `dst` register which is being set by `setb`.

Reply:
> @vamsi-parasa Other boolean producers such as `Conv2B` or `VectorTest` widen the byte value to 32 bit after `setb` so I believe a zeroing is necessary here. You accidentally achieve correct results in tests because the boolean stores only read the least significant 8 bits of the `dst` registers. Other operations such as `test` will read full `int` and may lead to incorrect results.
> 
> @jatin-bhateja Boolean is just dressed-up int except in stores so most readers will consume 32 bits of the register.
> 
> Thanks.

For explicit widening from byte to int compiler will emit an explicit zero extension instruction (movzbl).  There will not be any case where byte will be directly read along with unmodified bits  a consumer.